sdcflows/interfaces/bspline.py

```diff
@@ -0,0 +1,387 @@
+
+    This interface resolves the optimization problem of obtaining the B-Spline coefficients
+    :math:`c(\mathbf{k})` that best approximate the data samples within the
+    brain mask :math:`f(\mathbf{s})`, following Eq. (17) -- in that case for 2D --
+    of [Unser1999]_.
+    Here, and adapted to 3D:
+
+    .. math::
+
+        f(\mathbf{s}) =
+        \sum_{k_1} \sum_{k_2} \sum_{k_3} c(\mathbf{k}) \Psi^3(\mathbf{k}, \mathbf{s}).
+        \label{eq:1}\tag{1}
+
+    References
+    ----------
+    .. [Unser1999] M. Unser, "`Splines: A Perfect Fit for Signal and Image Processing
+        <http://bigwww.epfl.ch/publications/unser9902.pdf>`__," IEEE Signal Processing
+        Magazine 16(6):22-38, 1999.
+
+    See Also
+    --------
+    :py:func:`bspline_weights` - for Eq. :math:`\eqref{eq:2}` and the evaluation of
+    the tri-cubic B-Splines :math:`\Psi^3(\mathbf{k}, \mathbf{s})`.
+
+    """
+
+    input_spec = _BSplineApproxInputSpec
+    output_spec = _BSplineApproxOutputSpec
+
+    def _run_interface(self, runtime):
+        from sklearn import linear_model as lm
+
+        # Load in the fieldmap
+        fmapnii = nb.load(self.inputs.in_data)
+        data = fmapnii.get_fdata(dtype="float32")
+        mask = nb.load(self.inputs.in_mask).get_fdata() > 0
+        bs_spacing = [np.array(sp, dtype="float32") for sp in self.inputs.bs_spacing]
+
+        # Recenter the fieldmap
+        if self.inputs.recenter == "mode":
+            from scipy.stats import mode
+
+            data -= mode(data[mask], axis=None)[0][0]
+        elif self.inputs.recenter == "median":
+            data -= np.median(data[mask])
+        elif self.inputs.recenter == "mean":
+            data -= np.mean(data[mask])
+
+        # Calculate spatial location of voxels, and normalize per B-Spline grid
+        mask_indices = np.argwhere(mask)
+        fmap_points = apply_affine(fmapnii.affine.astype("float32"), mask_indices)
+
+        # Calculate the spatial location of control points
+        bs_levels = []
+        w_l = []
+        ncoeff = []
+        for sp in bs_spacing:
+            level = bspline_grid(fmapnii, control_zooms_mm=sp)
+            bs_levels.append(level)
+            ncoeff.append(level.dataobj.size)
+            w_l.append(bspline_weights(fmap_points, level))
+
+        # Compose the interpolation matrix
+        regressors = np.vstack(w_l)
+
+        # Fit the model
+        model = lm.Ridge(alpha=self.inputs.ridge_alpha, fit_intercept=False)
+        model.fit(regressors.T, data[mask])
+
+        interp_data = np.zeros_like(data)
+        interp_data[mask] = np.array(model.coef_) @ regressors  # Interpolation
+
+        # Store outputs
+        out_name = fname_presuffix(
+            self.inputs.in_data, suffix="_field", newpath=runtime.cwd
+        )
+        hdr = fmapnii.header.copy()
+        hdr.set_data_dtype("float32")
+        nb.Nifti1Image(interp_data, fmapnii.affine, hdr).to_filename(out_name)
```

Comment:
should allow nifti2 as well?
```suggestion
        fmapnii.__class__(interp_data, fmapnii.affine, hdr).to_filename(out_name)
```